packages/toolpad-core/src/Crud/FormPage.tsx

```diff
@@ -0,0 +1,376 @@
+'use client';
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import Box from '@mui/material/Box';
+import Button from '@mui/material/Button';
+import Checkbox from '@mui/material/Checkbox';
+import FormControl from '@mui/material/FormControl';
+import FormControlLabel from '@mui/material/FormControlLabel';
+import FormGroup from '@mui/material/FormGroup';
+import FormHelperText from '@mui/material/FormHelperText';
+import Grid from '@mui/material/Grid2';
+import InputLabel from '@mui/material/InputLabel';
+import MenuItem from '@mui/material/MenuItem';
+import Select, { SelectChangeEvent } from '@mui/material/Select';
+import TextField from '@mui/material/TextField';
+import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
+import { DatePicker } from '@mui/x-date-pickers/DatePicker';
+import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker';
+import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
+import type { GridSingleSelectColDef } from '@mui/x-data-grid';
+import dayjs, { Dayjs } from 'dayjs';
+import { useNotifications } from '../useNotifications';
+import { DataField, DataModel, DataSource, OmitId } from './shared';
+
+interface FormPageLocaleText {
+  submitButtonLabel: string;
+  submitSuccessMessage: string;
+  submitErrorMessage: string;
+}
+
+export interface FormPageProps<D extends DataModel> {
+  dataSource:
+    | (DataSource<D> & Required<Pick<DataSource<D>, 'createOne'>>)
+    | (DataSource<D> & Required<Pick<DataSource<D>, 'updateOne'>>);
+  initialValues?: Partial<OmitId<D>>;
+  onSubmit: (formValues: Partial<OmitId<D>>) => void | Promise<void>;
+  onSubmitSuccess?: () => void;
+  resetOnSubmit?: boolean;
+  localeText: FormPageLocaleText;
+}
+
+/**
+ * @ignore - internal component.
+ */
+function FormPage<D extends DataModel>(props: FormPageProps<D>) {
+  const {
+    dataSource,
+    initialValues = {} as Partial<OmitId<D>>,
+    onSubmit,
+    onSubmitSuccess,
+    resetOnSubmit = true,
+    localeText,
+  } = props;
+  const { fields, validate } = dataSource;
+
+  const notifications = useNotifications();
+
+  const [formState, setFormState] = React.useState<{
```

Comment:
Not using any external form libraries, just keeping form values in state.
It made the implementation more difficult and very specific + adds an extra choice for the user, this implementation seems to work well so far.

Reply:
Update: I've made the `CrudForm` component being used by the `Create` and `Edit` components implementation-agnostic and exported, so it should be possible to use it with any form library in advanced customization use cases.

https://github.com/mui/toolpad/pull/4486/files#diff-774ee88d140c8d65cc0b84644c1a6ae920994e79a17c3d751235e53f04bd8cdd